### DISCUSSION TITLE
Custom v-model modifiers

### DISCUSSION THREAD

**Discussion by jods4** [vuejs/rfcs#376] [General]
I wish that `v-model` (for native DOM elements) had extensible modifiers.

In other words: that there was an API to provide our own modifiers during app configuration and that `.trim`, `.number` and `.lazy` where things that were doable in userland.

Those three are nice, but it's a really small subset compared to all the useful things the community could dream.
On the other hand, because of size and maintainability, it's not in Vue interest to add more and more modifiers to its core lib.

It's the perfect thing to be provided by community, but it's not possible today.

### Use cases
I'm sure you can come up with lots of use-cases, here are just a few: 
`.null` replaces empty strings with `null` values
`.int` like `.number` but enforces integer values
`.number` like `.number` 😆 but with less error-tolerant parsing
`.date`, `.datetime`, `.time`, `.interval`, `.bool` and any type you can think of
`.upper` and `.lower` to change the casing of text
`.lazy` but with other events?
`.empty` displays a specific value when source is null or empty (both ways). 

Slight problem with the last one: this modifier would need an argument. TBD but since v-model has access to the DOM element, the argument could be in `data-empty` for example.

### Alternative 1: computed or custom ref
If you back your v-model with a computed or custom ref, you can do some transformations at this level.

This is a less convenient but more importantly, it is impossible except for the simplest scenarios.

`.trim` for example is a lot more subtle than you think.
If you trim the string in your model as the user types, it becomes impossible to type sentences with spaces, as they're removed as soon as you type them, even if you want to continue typing. Adding `.lazy` helps but prevents you from updating your UI as the user types.

If you look at the source of Vue, `v-model` contains code specifically to avoid updating the UI when `.trim` is used if it only differs from model by spaces. The UI is eventually trimmed in `@change`. 
This is good but impossible to replicate in a `computed`.

### Alternative 2: use functions in bindings
You can format anything with a function, and for convenience you can declare global functions available everywhere.
But they don't work in `v-model`, because the v-model expression must be writable.

### Alternative 3: wrap native elements
This is really heavy-weight if `.null` is all you wanted...
If you wrap elements then you can do whatever you want in `v-model`.

### Alternative 4: do everything in code
Wire-up the transformations you want in an additional layer everywhere you want it.
This is a lot of repeated boilerplate, esp. in large applications with lots of screens.

**Comment by CyberAP** [comment]
I would also like to have more advanced `v-model` cases with less boilerplate, but then we have these complex problems:

1. Should we be able to override buil-in modifiers?
2. How do we distinguish built-in modifiers from custom ones in that case?
3. Where do we define the logic for modifiers? Should it be global or local? How can we code-split logic for global modifiers? How do we re-use local logic?
4. Can we extend one modifier upon another?

This also raises all the problems from #188 when `v-model` doesn't actually control the value that is displayed to the user.

**Reply by jods4** [reply]
1. I'd say no, because it's simpler and not really limiting.
2. n/a because I said no to 1?
3. Again for simplicity I'm thinking: global. 
You register them at application startup like global components or directives.
How many and how large do you think modifiers can get? Does it warrant code-splitting?
4. Depends on what you mean? And how the API is shaped... Probably you can define a new modifier that is implemented by calling the functions of another modifier, which you want to extend/modify?

Number 3. makes me think of a potential problem when re-using components from libraries. 
If a component library uses its own modifiers, they will need to be globally registered. 
No big deal except there's a chance they'll collide with modifiers of another library :(

Global names are always bad... maybe it needs some namespacing for libraries, or maybe local modifiers are a good idea after all.
Not sure about this one 🤔

**Comment by jods4** [comment]
Friendly bump of this topic after the release of stable `defineModel`.

`defineModel` offers a new way to handle modifiers, by having configurable `get()` and `set(value)` functions associated with model.
`defineModel` also accepts a second generic argument that describes supported modifiers (for tooling, e.g. `defineModel<number, "round" | "abs">`).

It's not composable and it just doesn't feel "right" that the model modifiers must be hard-coded into each component models.
Especially when exposed through `get` and `set`, modifiers are bi-directional pipes that modify values (at their core, ideally they should be able to do more than that, see `.lazy` for example). These could be applied from outside, by SFC compiler.

There's no reason to hardcode lists of `if (mods.xxx)` in every model setter.
There's no reason I shouldn't be able to use another modifier that the component author did not think of (what if `capitalize` and `camelCase` are supported but not `snakeCase` or `kebabCase`?).

I think Vue should try to come up with a design for a Modifier API that can be applied to any `v-model`; and models should not care about implementing modifiers (separation of concerns).
